File: nim/day03/solve.nim
import std/[sugar, strformat, strutils, sequtils]

const BITS = 12
const MASK = (1 shl BITS) - 1

func parseBitstring(input: string): int =
  input.foldl(a shl 1 or int b != '0', 0)

func bitAt(bits: int, index: int): bool =
  (bits shr (BITS - index - 1) and 1) == 1

func mostCommonBit(nums: openArray[int], bitIndex: int): bool =
  let count = nums.countIt it.bitAt bitIndex
  2 * count >= nums.len

func gammaEpsilon(nums: openArray[int]): (int, int) =
  var gamma = 0
  for bitIndex in 0 ..< BITS:
    gamma = gamma shl 1 or int nums.mostCommonBit bitIndex
  (gamma, gamma xor MASK)

func findByCriterion(nums: openArray[int], byMostCommon: bool): int =
  var nums = nums.toSeq
  for bitIndex in 0 ..< BITS:
    let mostCommon = nums.mostCommonBit bitIndex
    let bit = mostCommon xor byMostCommon
    nums.keepIf bits => bit == bits.bitAt bitIndex
    if nums.len == 1:
      return nums[0]
  raise Exception.newException "ran out of oxygen lmao"

proc solve* =
  let nums = ".input/day03".readFile.strip.split.map parseBitstring
  let (gamma, epsilon) = nums.gammaEpsilon
  let oxygen = nums.findByCriterion true
  let co2 = nums.findByCriterion false
  let part1 = gamma * epsilon
  let part2 = oxygen * co2
  echo &"day03: {part1} {part2}"

File: kotlin/day03/solve.kt
package aoc.day03

import aoc.utils.*

internal const val BITS = 12
internal const val MASK = (1 shl BITS) - 1

internal fun Int.bitAt(index: Int) =
  shr (BITS - index - 1) and 1 == 1

internal fun Iterable<Int>.mostCommonBit(bitIndex: Int) =
  2 * count { it.bitAt(bitIndex) } >= count()

internal fun Iterable<Int>.gammaEpsilon(): Pair<Int, Int> {
  val gamma = (0 until BITS).asSequence()
    .map { bitIndex -> mostCommonBit(bitIndex).toInt() }
    .fold(0) { bits, bit -> bits shl 1 or bit }
  return Pair(gamma, gamma xor MASK)
}

internal fun Iterable<Int>.findByCriterion(byMostCommon: Boolean): Int {
  val nums = toMutableList()
  for (bitIndex in 0 until BITS) {
    val mostCommon = nums.mostCommonBit(bitIndex)
    val bit = mostCommon xor byMostCommon
    nums.retainAll { it.bitAt(bitIndex) == bit }
    if (nums.count() == 1)
      return nums.first()
  }
  unreachable()
}

fun solve() {
  val input = java.io.File(".input", "day03")
  val nums = input.useLines { it.map { it.toInt(2) }.toList() }
  val (gamma, epsilon) = nums.gammaEpsilon()
  val oxygen = nums.findByCriterion(true)
  val co2 = nums.findByCriterion(false)
  val part1 = gamma * epsilon
  val part2 = oxygen * co2
  println("day03: $part1 $part2")
}
